clang/lib/CodeGen/CGCall.cpp

```diff
@@ -5390,11 +5390,19 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
             V->getType()->isIntegerTy())
           V = Builder.CreateZExt(V, ArgInfo.getCoerceToType());
 
-        // If the argument doesn't match, perform a bitcast to coerce it.  This
-        // can happen due to trivial type mismatches.
+        // If the argument doesn't match, we are either trying to pass an
+        // alloca-ed sret argument directly, and the alloca AS does not match
+        // the default AS, case in which we AS cast it, or we have a trivial
+        // type mismatch, and thus perform a bitcast to coerce it.
         if (FirstIRArg < IRFuncTy->getNumParams() &&
-            V->getType() != IRFuncTy->getParamType(FirstIRArg))
-          V = Builder.CreateBitCast(V, IRFuncTy->getParamType(FirstIRArg));
+            V->getType() != IRFuncTy->getParamType(FirstIRArg)) {
+          auto IRTy = IRFuncTy->getParamType(FirstIRArg);
+          auto MaybeSRetArg = dyn_cast_or_null<llvm::Argument>(V);
+          if (MaybeSRetArg && MaybeSRetArg->hasStructRetAttr())
+            V = Builder.CreateAddrSpaceCast(V, IRTy);
```

Comment:
As a general rule, we try to use the target hook to perform address-space conversions.  That target hook is expressed in terms of AST address spaces, which is one reason I think we need to thread a LangAS through.  If we need to do the same for the other indirect cases, so be it.

Reply:
That's fair, but I'm not entirely sure that isn't simply excessive here - we already have the types, and the only mismatch for `sret` can be in the AS, I believe; reverting to LangAS from target ASes seems a bit roundabout. I think @arsenm had a related objection to this cast being unconditional, which I haven't handled.

Reply:
This is the prevailing existing practice in Clang CodeGen; you'll notice we do the same thing in `CreateTempAlloca`.  We are trying to allow targets to completely own the lowering of address spaces to IR.  The idea is that targets may want to distinguish address spaces in the frontend without distinguishing them in the backend, or they may decide that they need the address space conversion operation to be more complex than a simple IR `addrspacecast`.

Reply:
OK, I *think* that I've found a possibly acceptable middle ground (both for this and your other objection). Note that I am not rejecting the fact that we probably want `LangAS`es threaded through the indirect interfaces that deal with ASes, but it gives me a bit of trepidation to do it as part of this PR. I'd prefer to wrap this up, as it blocks some other work, and then open a separate PR/discussion around re-doing the interfaces.